src/core/backbone_router/multicast_listeners_table.cpp

```diff
@@ -0,0 +1,310 @@
+    LogMulticastListenersTable("Add", aAddress, aExpireTime, error);
+    CheckInvariants();
+    return error;
+}
+
+void MulticastListenersTable::Remove(const Ip6::Address &aAddress)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    for (uint16_t i = 0; i < mNumValidListeners; i++)
+    {
+        Listener &listener = mListeners[i];
+
+        if (listener.GetAddress() == aAddress)
+        {
+            mNumValidListeners--;
+
+            if (i != mNumValidListeners)
+            {
+                listener = mListeners[mNumValidListeners];
+                FixHeap(i);
+            }
+
+            if (mCallback != nullptr)
+            {
+                mCallback(mCallbackContext, OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED, &aAddress);
+            }
+
+            ExitNow(error = OT_ERROR_NONE);
+        }
+    }
+
+exit:
+    LogMulticastListenersTable("Remove", aAddress, static_cast<TimeMilli>(0), error);
+    CheckInvariants();
+}
+
+void MulticastListenersTable::Expire(void)
+{
+    TimeMilli    now = TimerMilli::GetNow();
+    Ip6::Address address;
+
+    while (mNumValidListeners > 0 && now >= mListeners[0].GetExpireTime())
+    {
+        LogMulticastListenersTable("Expire", mListeners[0].GetAddress(), mListeners[0].GetExpireTime(), OT_ERROR_NONE);
+        address = mListeners[0].GetAddress();
+
+        mNumValidListeners--;
+
+        if (mNumValidListeners > 0)
+        {
+            mListeners[0] = mListeners[mNumValidListeners];
+            FixHeap(0);
+        }
+
+        if (mCallback != nullptr)
+        {
+            mCallback(mCallbackContext, OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED, &address);
+        }
+    }
+
+    CheckInvariants();
+}
+
+void MulticastListenersTable::LogMulticastListenersTable(const char *        aAction,
+                                                         const Ip6::Address &aAddress,
+                                                         TimeMilli           aExpireTime,
+                                                         otError             aError)
+{
+    OT_UNUSED_VARIABLE(aAction);
+    OT_UNUSED_VARIABLE(aAddress);
+    OT_UNUSED_VARIABLE(aExpireTime);
+    OT_UNUSED_VARIABLE(aError);
+
+    otLogDebgBbr("MulticastListenersTable: %s %s expire %lu: %s", aAction, aAddress.ToString().AsCString(), aExpireTime,
+                 otThreadErrorToString(aError));
+}
+
+void MulticastListenersTable::FixHeap(uint16_t aIndex)
+{
+    if (!SiftHeapElemDown(aIndex))
+    {
+        SiftHeapElemUp(aIndex);
+    }
+}
+
+void MulticastListenersTable::CheckInvariants(void) const
+{
+#if OPENTHREAD_EXAMPLES_SIMULATION
+    OT_ASSERT(std::is_heap(&mListeners[0], &mListeners[mNumValidListeners], Listener::IsAfter));
+#endif
+}
+
+bool MulticastListenersTable::SiftHeapElemDown(uint16_t aIndex)
+{
+    uint16_t i = aIndex;
+    Listener saveElem;
+
+    OT_ASSERT(aIndex < mNumValidListeners);
+
+    saveElem = mListeners[aIndex];
+
+    for (;;)
+    {
+        uint16_t j = 2 * i + 1;
```

Comment:
how about calling it `left = 2 * index + 1` and `right = 2 * index + 2` (for left and right node indexes of binary tree)?

Reply:
renamed `j` to `child`. Didn't use `left` because `child` might be pointing to the right child.